### ISSUE TITLE
PageContext list & listItem are undefined until I refresh the page

### ISSUE THREAD

**Issue by Misuki-CG** [SharePoint/sp-dev-docs#3977]
#### Category
- [x] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

I want to get information about the current page my webpart is installed on. The page itself is not already publish, but it can be saved as a draft or something like that. 
I searched for these informations and I came to the point that it can be a bug. I am using `this.context.pageContext.list` and `this.context.pageContext.listItem` to retrieve the name of the aspx file associated with the current page. 

#### Observed Behavior

My problem is that these fields (`list`& `listItem` in pageContext are always "undefined" until i'm refreshing the page (with F5 for example). It happens even if the page is already published and I install the WebPart. 

#### Steps to Reproduce

Create a new modern sharepoint online page, edit the title, then add a webpart. If you log this.context.pageContext in the onInit() method of the WebPart for example, you'll see that `listItem` and `list` are undefined. Even if you pass the context as a props to the react component and then you log it in the componentDidMount() / componentDidUpdate() function. 

So the question is : is this a bug ? is there any other way to get this information ? 

Thank you. 

#### Other information

I am using yoman to generate the webpart source files. I use ReactJS & TypeScript. I use pnp sp. I am currently on the 1.8.0 version. 

Ff there is lack of information or something like that, feel free to tell me and I'll answer you.

**Comment by andrewconnell** [comment]
Can you provide additional context information, such as SharePoint environment (SP16/SP19/SPO), browser, etc?

Have you also tried using the latest SPFx version (1.8.2)?

Also, can you share the relevant code in your web part, showing where you are referencing the context?

**Comment by Misuki-CG** [comment]
I am using Sharepoint Online & my browser is Google Chrome (in fact, I tested it with Firefox & Google Chrome, last version for both). 

At this moment I'm not able to upgrade to SPFx version 1.8.2 (this was my first request to my superior). 

I'm currently referencing the context in the WebPart onInit() function : 

```
import { sp } from "@pnp/sp"; 

  public onInit(): Promise<void>{
    return super.onInit().then(() => {
      sp.setup({
        spfxContext: this.context // because i am using pnpsp library
      });

      console.log(this.context.pageContext); // here it is
    });
  }
```

Full log of pageContext : 

![image](https://user-images.githubusercontent.com/23389135/58075844-ff551900-7ba8-11e9-8d23-94b3bb561fc3.png)

Note: in legacyPageContext, it's the same : all informations related to the page itself are undefined or empty.

**Comment by Misuki-CG** [comment]
Update :

in the same environment as above, I find out that there are parameters in the URL of the webpart ( accessible by `window.location.href`). When the webpart is installing, there is no ID or LIST parameter in the URL . However, in any other cases where the webpart is already installed, they are here. 

How does the installation of the webpart works ? it may forget to put these parameters in the URL ?

**Comment by clechner77** [comment]
@Misuki-CG  This should get you pretty far..........

`
import ITesterProps from './components/ITester/ITesterProps'

export interface ITesterWebPartProps {
    description:string;
    pageProps:Promise<any>;
}

export default class TesterWebPart extends BaseClientSideWebPart<ITesterWebPartProps> {
  public onInit(): Promise<void> {
    return super.onInit().then(_ => {
      pnpSetup({ spfxContext: this.context });
    });
  }

  public getPageProps():Promise<any>{
    let _list = this.context.pageContext.list ? this.context.pageContext.list.title : null;
    let _id = this.context.pageContext.listItem ? this.context.pageContext.listItem.id : null;
    if (_list && _id){
      return sp.web.lists.getByTitle(`${_list}`)
      .items.getById(_id)
      .get().then((page)=>{
        console.log(page);
        return page
      });
    }
  }

  public render(): void {
    const element: React.ReactElement<ITesterProps > = React.createElement(
      Tester,
      {
        description: this.description,
        pageProps: this.getPageProps(),
      },
    );
    ReactDom.render(element, this.domElement);
  }
  protected onDispose(): void {
    ReactDom.unmountComponentAtNode(this.domElement);
  }
  protected get dataVersion(): Version {
    return Version.parse('1.0');
  }
  protected getPropertyPaneConfiguration(): IPropertyPaneConfiguration {
    return {
      pages: [
        {
          header: {
            description: strings.PropertyPaneDescription
          },
          groups: [
            {
              groupName: strings.BasicGroupName,
              groupFields: [
                PropertyPaneTextField('description', {
                  label: strings.DescriptionFieldLabel
                }), 
              ],
            }
          ]
        }
      ]
    };
  }
}


`

**Comment by Misuki-CG** [comment]
@clechner77 Thank you for your reply. 

I see what your solution was about, but it does not change the fact that `list` and `listItem` are undefined at the start of the webpart install process. I actually implemented the code you wrote in another way until it's fixed but it force the customer to refresh the page when he installs the webpart & that's a problem atm. 

Is there any other way to get current page information ?

**Comment by clechner77** [comment]
So, is the context initially available, and then after inserting the webpart and saving it, the context disappears?
________________________________

**Comment by Misuki-CG** [comment]
@clechner77 All the context, except for the fields `context.pageContext.list`, `context.pageContext.listItem` and `context.pageContext.page`, is available no matter what. 

After inserting the webpart (so when we click on the "+" and add the webpart to the page), the fields
`context.pageContext.list`, `context.pageContext.listItem` and `context.pageContext.page` are `undefined`. 

After inserting the webpart AND refreshing the page (with F5 for example), all the context is available. 

The problem is that I need to have the page ID (via `context.pageContext.listItem`) at the moment where I insert the webpart in the page, so my webpart can work correctly.

**Comment by clechner77** [comment]
Part of the issue is that there is no page context when it's a new page, so I'm guessing for uniformity they elected to not make it available immediately when editing a page as well.  In your case, you would have to deal with that problem if someone created a new page right?
________________________________

**Comment by Misuki-CG** [comment]
Yep, that's right. 

I tried to do it on an existing page : it is the same. I installed the webpart in an existing page and I had no page context. 

But you're right, I want to deal with this when someone created a new page. 

I'm pretty sure that the ID of the listItem associated to the current page already exist when we are creating the page, because the list item has a random name (like "dsfjhHD.aspx") but when you set the title of the page and hit "save as a draft", it makes an API call on `_api/sitepages/pages(**ID of the current page**)/SavePageAsDraft`. 

I think it's a provisionning bug like, the item ID of the page or/and the list ID is not given to the webpart so it can't construct the page context, or something like this (I don't really know how the context is passing to the WebPart).
